### ISSUE TITLE
Constructors should not have side effects

### ISSUE THREAD

**Issue by Corniel** [SonarSource/sonar-dotnet#2231]
### Description
A constructor should initialize the instance of the class it belongs to, but nothing else.

### Expected behavior
``` C#
public class MyClass
{
    public MyClass(SomeClass input)
    {
        this.Prop = input.Prop // Compliant
        input.Count++; // Not compliant as it changes the state
        this.Calculated = input.Calculate(); // Compliant. At least, unless we can check the internals there
        input.Update(this); // Not compliant. the method call does not change the state of the construced

       SetSomeStuff(input) // Not compliant. SetSomeStuff changes the state of input.
    }
    private void SeSomeStuff(input)
    {
        input.Done = true;
    }
}
```

**Comment by Evangelink** [comment]
Hi @Corniel,

I assume that you are suggesting a new rule and I will thank you for that. 

I am not entirely sure of the scope of what you would want to detect, is it all setters and all void methods on an object passed as an argument of the constructor? 

Do you think about also raising an issue for objects created inside the constructor? 

Shall we also raise an issue for well-known (system) classes, for example, if you take a `HttpClient` object and you set something (like the header)?

**Comment by Corniel** [comment]
Creating new objects that become part of the class should be fine. But changing the state of passed parameters is something that should not happen.

Typically something that go wrong:

``` C#
var x = new CustomObjectWithSideEffect(someInput);
```
And that x is not longer used, and therefore removed, but because of the side effects breaks the code. On top of that, it is something that is hard to understand from the intent of your code.

**Comment by Corniel** [comment]
@Evangelink Allready gave this some extra thoughts?

**Comment by Evangelink** [comment]
Hi @Corniel,

Sorry not to have returned to you before. I do think this is an interesting idea and I have some idea about how to implement it but I am afraid it won't be done soon.

**Comment by Corniel** [comment]
@Evangelink I also have ideas on how it can/should be implemented. Given your backlog I understand that it will take some time, but I believe this is really valuable rule, so I hope it will be done eventually.

**Comment by pavel-mikula-sonarsource** [comment]
Internal ticket [NET-1969](https://sonarsource.atlassian.net/browse/NET-1969)

[NET-1969]: https://sonarsource.atlassian.net/browse/NET-1969?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ
